module/validation/receipt_validator.go

```diff
@@ -148,29 +149,152 @@ func (v *receiptValidator) resultChainCheck(result *flow.ExecutionResult, prevRe
+func (v *receiptValidator) Validate(receipt *flow.ExecutionReceipt) error {
+	// TODO: this can be optimized by checking if result was already stored and validated.
+	// This needs to be addressed later since many tests depend on this behavior.
+	prevResult, err := v.fetchResult(receipt.ExecutionResult.PreviousResultID)
+	if err != nil {
+		return err
+	}
+
+	// first validate result to avoid signature check in in `validateReceipt` in case result is invalid.
+	err = v.validateResult(&receipt.ExecutionResult, prevResult)
+	if err != nil {
+		return fmt.Errorf("could not validate single result %v at index: %w", receipt.ExecutionResult.ID(), err)
+	}
+
+	err = v.validateReceipt(receipt.Meta(), receipt.ExecutionResult.BlockID)
+	if err != nil {
+		// It's very important that we fail the whole validation if one of the receipts is invalid.
+		// It allows us to make assumptions as stated in previous comment.
+		return fmt.Errorf("could not validate single receipt %v: %w", receipt.ID(), err)
+	}
+
+	return nil
+}
+
+func (v *receiptValidator) ValidatePayload(candidate *flow.Block) error {
 	// lookup cache to avoid linear search when checking for previous result that is
 	// part of payload
-	payloadExecutionResults := make(map[flow.Identifier]*flow.ExecutionResult)
-	for _, receipt := range receipts {
-		payloadExecutionResults[receipt.ExecutionResult.ID()] = &receipt.ExecutionResult
-	}
+	payloadExecutionResults := candidate.Payload.ResultsById()
 	// Build a functor that performs lookup first in receipts that were passed as payload and only then in
 	// local storage. This is needed to handle a case when same block payload contains receipts that
 	// reference each other.
 	// ATTENTION: Here we assume that ER is valid, this lookup can return a result which is actually invalid.
 	// Eventually invalid result will be detected and fail the whole validation.
-	previousResult := func(executionResult *flow.ExecutionResult) (*flow.ExecutionResult, error) {
-		prevResult, found := payloadExecutionResults[executionResult.PreviousResultID]
+	fetchResult := func(previousResultID flow.Identifier) (*flow.ExecutionResult, error) {
+		prevResult, found := payloadExecutionResults[previousResultID]
 		if found {
 			return prevResult, nil
 		}
 
-		return v.previousResult(executionResult)
+		return v.fetchResult(previousResultID)
+	}
+
+	header := candidate.Header
+	payload := candidate.Payload
+
+	// Get the latest sealed block on this fork, ie the highest block for which
+	// there is a seal in this fork. This block is not necessarily finalized.
+	last, err := v.seals.ByBlockID(header.ParentID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve parent seal (%x): %w", header.ParentID, err)
 	}
+	sealed, err := v.headers.ByBlockID(last.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve sealed block (%x): %w", last.BlockID, err)
+	}
+	sealedHeight := sealed.Height
```

Comment:
we could get the sealed height from protocol state: `state.Sealed().Head().Height`

Reply:
I don't think it's a valid change though, we need to check fork wise, so it doesn't work.

Reply:
right. `state.Sealed().Head()` gives you sealed and _finalized_, whereas your current solution doesn't require finalized.